ansible/roles/fastpath/templates/ooni_api_uploader.py

```diff
@@ -0,0 +1,228 @@
+                except Exception:
+                    pass
+
+            if msm is None:
+                log.error("Unable to parse measurement")
+                jf.write(b"{}\n")
+                continue
+
+            jf.write(ujson.dumps(msm).encode())
+            jf.write(b"\n")
+
+            rid = msm.get("report_id") or ""
+            input = msm.get("input") or ""
+            msmt_uid = msmt_f.name[:-5]
+            d = dict(
+                report_id=rid, input=input, measurement_uid=msmt_uid, linenum=linenum
+            )
+            lookup_list.append(d)
+
+    return lookup_list
+
+
+def delete_msmt_posts(measurements: List[PP]) -> None:
+    log.info(f"Deleting {len(measurements)} measurements")
+    for msmt_f in measurements:
+        msmt_f.unlink()
+
+
+@metrics.timer("total_run_time")
+def main():
+    conf = read_conf()
+    bucket_name = conf.get("bucket_name")
+    spooldir = Path(conf.get("msmt_spool_dir"))
+    format_char = "n"
+    collector_id = conf.get("collector_id")
+    identity = f"{format_char}{collector_id}"
+    log.info(f"Uploader {collector_id} starting")
+    assert collector_id, "collector_id is not set"
+    log.info(f"Using bucket {bucket_name} and spool {spooldir}")
+
+    s3 = create_s3_client(conf)
+
+    db_conn = connect_to_db(conf)
+
+    # Scan spool directories, by age
+    idir = spooldir / "incoming"
+    threshold = datetime.utcnow() - timedelta(hours=1)
+    for hourdir in sorted(idir.iterdir()):
+        if not hourdir.is_dir() or hourdir.suffix == ".tmp":
+            continue
+        try:
+            tstamp, cc, testname = hourdir.name.split("_")
+        except Exception:
+            continue
+        if len(tstamp) != 10:
+            continue
+        hourdir_time = datetime.strptime(tstamp, "%Y%m%d%H")
+        if hourdir_time > threshold:
+            log.info(f"Stopping before {hourdir_time}")
+            break
+
+        log.info(f"Processing {hourdir}")
+        # Split msmts across multiple postcans and jsonl files
+        can_cnt = 0
+        while True:
+            # Compress raw POSTs into a tar.gz postcan
+            postcanf = hourdir.with_suffix(f".{identity}.{can_cnt}.tar.gz")
+            jsonlf = hourdir.with_suffix(f".{identity}.{can_cnt}.jsonl.gz")
+            msmfiles = fill_postcan(hourdir, postcanf)
+            if len(msmfiles) == 0:
+                break
+            # Also create jsonl file and delete msmt POSTs
+            lookup_list = fill_jsonl(msmfiles, jsonlf)
+            delete_msmt_posts(msmfiles)
+
+            # Upload current postcan to S3
+            postcan_s3path = (
+                f"raw/{tstamp[:8]}/{tstamp[8:10]}/{cc}/{testname}/{postcanf.name}"
+            )
+            jsonl_s3path = (
+                f"raw/{tstamp[:8]}/{tstamp[8:10]}/{cc}/{testname}/{jsonlf.name}"
+            )
+            if conf.get("run_mode", "") == "DESTROY_DATA":
+                log.info("Testbed mode: Destroying postcans!")
```

Comment:
This mode of operation scares me a bit. I would suggest we maybe remove it to avoid mistakes.